src/c2goto/library/pthread_lib.c

```diff
@@ -483,23 +582,90 @@ __ESBMC_HIDE:;
   return 0;
 }
 
+// The pthread_key_create() function shall create a thread-specific
+// data key visible to all threads in the process.
+// source: https://linux.die.net/man/3/pthread_key_create
 int pthread_key_create(pthread_key_t *key, void (*destructor)(void *))
 {
 __ESBMC_HIDE:;
-  // TODO
-  return 0;
+  __ESBMC_atomic_begin();
+  int result = 0;
+  __ESBMC_assert(
+    key != NULL,
+    "In pthread_key_create, key parameter must be different than NULL.");
+  // the value NULL shall be associated with the new key in all active threads
+  result = insert_key_value(mylist, __ESBMC_next_thread_key, NULL);
+  __ESBMC_thread_key_destructors[__ESBMC_next_thread_key] = destructor;
+  // store the newly created key value at *key
+  *key = __ESBMC_next_thread_key++;
+  if(result < 0)
+  {
+    if(nondet_bool())
+    {
+      // Insufficient memory exists to create the key.
+      result = ENOMEM;
+    }
+    else
+    {
+      // The system lacked the necessary resources
+      // to create another thread-specific data key, or
+      // the system-imposed limit on the total number of
+      // keys per process {PTHREAD_KEYS_MAX} has been exceeded.
+      result = EAGAIN;
+    }
+  }
+  __ESBMC_atomic_end();
+  return result;
 }
 
+// The pthread_getspecific() function shall return
+// the value currently bound to the specified key
+// on behalf of the calling thread.
+// source: https://linux.die.net/man/3/pthread_getspecific
 void *pthread_getspecific(pthread_key_t key)
 {
 __ESBMC_HIDE:;
-  // TODO
-  return NULL;
+  __ESBMC_atomic_begin();
+  void *result = NULL;
+  if(key <= __ESBMC_next_thread_key)
+  {
+    // Return the thread-specific data value associated
+    // with the given key.
+    mylist = head;
+    result = search_key(mylist, key);
+  }
+  else if(key > __ESBMC_next_thread_key)
+  {
+    // If no thread-specific data value is associated with key,
+    // then the value NULL shall be returned.
+    result = NULL;
+  }
```

Comment:
Correct me if I'm wrong, but this branch is completely irrelevant here. if `(key <= __ESBMC_next_thread_key)` is not true, key can only be `> __ESBMC_next_thread_key` and result is already null.

Reply:
Fixed.